File: barkr/models/media.py
"""
Module to implement a generic representation of media objects
that can be posted or retrieved as part of messages
"""

from dataclasses import dataclass
from typing import Final

SUPPORTED_MIME_TYPES: Final[set[str]] = {
    "image/jpeg",
    "image/png",
    "image/gif",
    "image/webp",
    "video/mp4",
    "video/quicktime",
}


@dataclass(frozen=True)
class Media:
    """
    A generic media object with a MIME type and the content
    in bytes.

    Can be retrieved as part of a message or posted
    to a social media network.
    """

    # Required attributes to represent a media object
    mime_type: str
    content: bytes

    # Optional metadata
    alt_text: str = ""

    def is_valid(self) -> bool:
        """
        Check if the media object is valid and can be
        posted to a social media network or otherwise
        handled.

        For a media object to be valid, it must have
        a non-empty content and a valid MIME type.
        """

        has_content = len(self.content) > 0
        has_mime_type = self.mime_type.strip()

        if not has_content or not has_mime_type:
            return False

        return self.mime_type in SUPPORTED_MIME_TYPES

File: models/media.go
package models

import "slices"

// supportedMIMETypes lists the MIME types that Barker can process for media attachments in messages.
var supportedMIMETypes = []string{
	"image/jpeg",
	"image/png",
	"image/gif",
	"image/webp",
	"video/mp4",
	"video/quicktime",
}

// Media represents a generic media attachment in a message, like an image or a video.
// The struct preserves the MIME type and binary content, as well as any alternative text,
// and allows the media to be retrieved as part of a message's content, or posted to a
// target platform.
type Media struct {
	MIMEType string // MIME type of the media (e.g., "image/png", "video/mp4").
	Content  []byte // binary content of the media.
	AltText  string // alternative text for the media, useful for accessibility and descriptions.
}

// IsValid reports whether the [Media] instance represents a valid, non-empty media attachment.
// Valid MIME types are enumerated in [supportedMIMETypes].
func (m *Media) IsValid() bool {
	if m == nil {
		return false
	}

	if m.MIMEType == "" || len(m.Content) == 0 {
		return false
	}

	return slices.Contains(supportedMIMETypes, m.MIMEType)
}
